src/Types/PlutusAddress.purs

```diff
@@ -0,0 +1,192 @@
+
+instance FromData StakingCredential where
+  fromData (Constr zero [ pd ]) =
+    StakingHash <$> fromData pd
+  fromData (Constr one [ slotD, txIxD, certIxD ]) =
+    StakingPtr <$> ({ slot: _, txIx: _, certIx: _} <$>
+      fromData slotD <*> fromData txIxD <*> fromData certIxD)
+  fromData _ = Nothing
+
+--------------------------------------------------------------------------------
+-- Address
+--------------------------------------------------------------------------------
+
+data PlutusAddress = Address
+  { addressCredential :: Credential
+  , addressStakingCredential :: Maybe StakingCredential
+  }
+
+derive instance Generic PlutusAddress _
+
+instance Show PlutusAddress where
+  show = genericShow
+
+instance ToData PlutusAddress where
+  toData (Address a) =
+    Constr zero [ toData a.addressCredential, toData a.addressStakingCredential ]
+
+instance FromData PlutusAddress where
+  fromData (Constr zero [ credD, stakingCredD ]) =
+    Address <$> ({ addressCredential: _, addressStakingCredential: _ } <$>
+      fromData credD <*> fromData stakingCredD)
+  fromData _ = Nothing
+
+--------------------------------------------------------------------------------
+-- IsoNativeForeign
+--------------------------------------------------------------------------------
+
+class IsoNativeForeign native foreign' | native -> foreign', foreign' -> native where
+  toNativeType :: foreign' -> Maybe native
+-- TODO: toForeignType :: native -> foreign'
+
+instance IsoNativeForeign PlutusAddress Serialization.Address where
+  toNativeType addrForeign = addrType >>= \addrType' ->
+    case toInt addrType' of -- TODO: Byron addresses
+      -- %b0000 | network tag | key hash | key hash
+      0 -> buildAddress pubKeyCredential $ Just $
+        map StakingHash <<< pubKeyCredential
+
+      -- %b0001 | network tag | script hash | key hash
+      1 -> buildAddress scriptCredential $ Just $
+        map StakingHash <<< pubKeyCredential
+
+      -- %b0010 | network tag | key hash | script hash
+      2 -> buildAddress pubKeyCredential $ Just $
+        map StakingHash <<< scriptCredential
+
+      -- %b0011 | network tag | script hash | script hash
+      3 -> buildAddress scriptCredential $ Just $
+        map StakingHash <<< scriptCredential
+
+      -- %b0100 | network tag | key hash | pointer
+      4 -> buildAddress pubKeyCredential $ Just stakingPtr
+
+      -- %b0101 | network tag | script hash | pointer
+      5 -> buildAddress scriptCredential $ Just stakingPtr
+
+      -- %b0110 | network tag | key hash
+      6 -> buildAddress pubKeyCredential Nothing
+
+      -- %b0111 | network tag | script hash
+      7 -> buildAddress scriptCredential Nothing
+
+      _ -> Nothing
+    where
+    addrBytes :: Array Int
+    addrBytes = byteArrayToIntArray $ addressBytes addrForeign
+
+    addrType :: Maybe UInt
+    addrType = head addrBytes >>= (pure <<< flip zshr (fromInt 4) <<< fromInt)
```

Comment:
Could you make some of these numbers constants, or add some documentation explaining their significance? This applies to the literals below as well